File: src/encoder.js
/**
 * AICL Encoder — trie-accelerated
 * 1. Longest-match via trie O(maxLen) per position
 * 2. Word-based fallback when bestLen==1 (with all-caps direct lookup)
 * 3. Literal
 */
import { build } from './dict.js';
import { codePoints, isPuaCodePoint, ESCAPE_MARKER, requireText } from './unicode.js';

function buildTrie(patternToSymbol){
  const root={children:new Map(), symbol:null};
  for(const [pat,sym] of patternToSymbol){
    if(pat.startsWith('MOD_')) continue;
    let node=root;
    for(const ch of pat){
      if(!node.children.has(ch)) node.children.set(ch,{children:new Map(), symbol:null});
      node=node.children.get(ch);
    }
    node.symbol=sym;
  }
  return root;
}
function trieMatch(root, chars, i){
  let node=root, bestLen=0, bestSym=null;
  for(let j=i;j<chars.length;j++){
    const nxt=node.children.get(chars[j]);
    if(!nxt) break;
    node=nxt;
    if(node.symbol){ bestLen=j-i+1; bestSym=node.symbol; }
  }
  return [bestLen, bestSym];
}
function extractWord(chars,i){
  if(i>=chars.length) return '';
  if(!/[a-zA-Z0-9_'-]/.test(chars[i])) return '';
  let w=chars[i];
  for(let j=i+1;j<chars.length;j++){
    if(/[a-zA-Z0-9_'-]/.test(chars[j])) w+=chars[j]; else break;
  }
  return w;
}
function punctuationToModifier(ch){
  const m={' ':'MOD_TRAIL_SPACE',',':'MOD_TRAIL_COMMA','.':'MOD_TRAIL_PERIOD','?':'MOD_TRAIL_QUESTION','!':'MOD_TRAIL_EXCL',';':'MOD_TRAIL_SEMI',':':'MOD_TRAIL_COLON',')':'MOD_TRAIL_RPAREN',']':'MOD_TRAIL_RBRACKET','}':'MOD_TRAIL_RBRACE','"':'MOD_TRAIL_RQUOTE'};
  return m[ch]||null;
}
function charCount(s){ return [...s].length; }

let trieCache=null;
function getTrie(patternToSymbol){
  if(trieCache) return trieCache;
  trieCache=buildTrie(patternToSymbol);
  return trieCache;
}
let fragmentCache=null;
function getFragmentPatterns(sortedPatterns){
  if(fragmentCache) return fragmentCache;
  fragmentCache = sortedPatterns.filter(p=> p.length>=2 && p.length<=4).sort((a,b)=> b.length - a.length || (a<b?-1:1));
  return fragmentCache;
}

export function encode(text, opts={}){
  requireText(text, 'text');
  const {sortedPatterns, patternToSymbol}=build();
  const steps=opts.steps?[]:null;
  const trackMapping=opts.trackMapping;
  const trie=getTrie(patternToSymbol);
  let output='',matches=0,literals=0;
  const chars=codePoints(text);
  // rawToAicl[pos] = index in AICL output where this raw char's symbol starts
  const rawToAicl = trackMapping ? new Array(chars.length).fill(-1) : null;
  let aiclPos=0;
  let i=0;
  while(i<chars.length){
    const [bestLen,bestSym]=trieMatch(trie, chars, i);
    let bestPat=null;
    if(bestSym) bestPat=chars.slice(i,i+bestLen).join('');

    if(bestLen===1){
      const word=extractWord(chars,i);
      if(word.length>1){
        const lower=word.toLowerCase();
        const isAllCaps = word===word.toUpperCase() && /[A-Z]/.test(word);
        if(isAllCaps){
          const capsSym=patternToSymbol.get(word);
          if(capsSym){
            if(trackMapping) for(let j=i;j<i+word.length;j++) rawToAicl[j]=aiclPos;
            output+=capsSym; matches++; if(steps) steps.push({type:'base',pattern:word,pos:i});
            aiclPos++;
            i+=word.length;
            while(i<chars.length){
              const mn=punctuationToModifier(chars[i]);
              if(mn){ const ms=patternToSymbol.get(mn); if(ms){ output+=ms; matches++; if(steps) steps.push({type:'modifier',name:mn,pos:i}); aiclPos++; i++; continue; } }
              break;
            }
            continue;
          }
          // No dedicated uppercase entry: reuse the lowercase base symbol and
          // let MOD_ALLCAPS transform the whole buffer — one modifier instead
          // of spelling the word letter-by-letter (which expanded it).
          const baseSym=patternToSymbol.get(lower);
          if(baseSym){
            if(trackMapping) for(let j=i;j<i+word.length;j++) rawToAicl[j]=aiclPos;
            output+=baseSym; matches++; if(steps) steps.push({type:'base',pattern:lower,pos:i});
            aiclPos++;
            const acs=patternToSymbol.get('MOD_ALLCAPS');
            if(acs){ output+=acs; matches++; if(steps) steps.push({type:'modifier',name:'MOD_ALLCAPS',pos:i}); aiclPos++; }
            i+=word.length;
            while(i<chars.length){
              const mn=punctuationToModifier(chars[i]);
              if(mn){ const ms=patternToSymbol.get(mn); if(ms){ output+=ms; matches++; if(steps) steps.push({type:'modifier',name:mn,pos:i}); aiclPos++; i++; continue; } }
              break;
            }
            continue;
          }
          // No base either: fall through to the shared fragment path; each
          // fragment rides MOD_ALLCAPS below.
        } else {
          // A whole-word base symbol expands to the lowercase pattern and
          // MOD_CAPS can only fix up the first char, so words with internal
          // capitals (camelCase) must not use the base-symbol path. The
          // fragment path below can capitalize each fragment individually.
          const rest=word.slice(1);
          if(rest===rest.toLowerCase()){
            const baseSym=patternToSymbol.get(lower);
            if(baseSym){
              if(trackMapping) for(let j=i;j<i+lower.length;j++) rawToAicl[j]=aiclPos;
              output+=baseSym; matches++; if(steps) steps.push({type:'base',pattern:lower,pos:i});
              aiclPos++;
              i+=lower.length;
              if(word[0]!==lower[0]){
                const cs=patternToSymbol.get('MOD_CAPS'); if(cs){ output+=cs; matches++; if(steps) steps.push({type:'modifier',name:'MOD_CAPS',pos:i}); aiclPos++; }
              }
              while(i<chars.length){
                const mn=punctuationToModifier(chars[i]);
                if(mn){ const ms=patternToSymbol.get(mn); if(ms){ output+=ms; matches++; if(steps) steps.push({type:'modifier',name:mn,pos:i}); aiclPos++; i++; continue; } }
                break;
              }
              continue;
            }
          }
        }
        let fragI=0, matchedFrag=false;
        const fragPatterns = getFragmentPatterns(sortedPatterns);
        while(fragI<lower.length){
          let bestFrag=null, bestFragLen=0;
          for(const frag of fragPatterns){
            if(frag.length <= lower.length - fragI && lower.startsWith(frag, fragI)){
              // For all-caps words MOD_ALLCAPS fixes every char of the
              // fragment, so the interior-capital guard doesn't apply.
              // Otherwise only usable when no capital falls INSIDE the span:
              // MOD_CAPS can capitalize the fragment's first char only.
              if(isAllCaps || word.slice(fragI+1, fragI+frag.length) === lower.slice(fragI+1, fragI+frag.length)){
                bestFrag=frag; bestFragLen=frag.length; break;
              }
            }
          }
          if(bestFrag){
            if(trackMapping) for(let j=i+fragI;j<i+fragI+bestFragLen;j++) rawToAicl[j]=aiclPos;
            const fs=patternToSymbol.get(bestFrag);
            if(fs){
              output+=fs; matches++; if(steps) steps.push({type:'fragment',pattern:bestFrag,pos:i+fragI}); aiclPos++;
              // All-caps words may hide capitals anywhere inside a fragment
              // (e.g. 4RC2 → frags "4r","c2"), so every letter-bearing
              // fragment needs MOD_ALLCAPS — its buffer is just that fragment.
              // Otherwise only the fragment's first char can differ (MOD_CAPS).
              const needsCaps = isAllCaps ? /[a-z]/.test(bestFrag) : word[fragI]!==lower[fragI];
              if(needsCaps){
                const cs=patternToSymbol.get(isAllCaps ? 'MOD_ALLCAPS' : 'MOD_CAPS');
                if(cs){ output+=cs; matches++; if(steps) steps.push({type:'modifier',name:isAllCaps?'MOD_ALLCAPS':'MOD_CAPS',pos:i+fragI}); aiclPos++; }
              }
              fragI+=bestFragLen; matchedFrag=true; continue;
            }
          }
          break;
        }
        if(matchedFrag && fragI>0){
          i+=fragI;
          while(i<chars.length){
            const mn=punctuationToModifier(chars[i]);
            if(mn){ const ms=patternToSymbol.get(mn); if(ms){ output+=ms; matches++; if(steps) steps.push({type:'modifier',name:mn,pos:i}); aiclPos++; i++; continue; } }
            break;
          }
          continue;
        }
      }
    }
    if(bestSym){
      if(trackMapping) for(let j=i;j<i+bestLen;j++) rawToAicl[j]=aiclPos;
      output+=bestSym; matches++; if(steps) steps.push({type:'match',pattern:bestPat,pos:i});
      aiclPos+=bestLen > 1 ? [...bestSym].length : 1;
      i+=bestLen; continue;
    }
    const ch=chars[i];
    if(trackMapping) rawToAicl[i]=aiclPos;
    output+= isPuaCodePoint(ch.codePointAt(0)) ? ESCAPE_MARKER+ch : ch;
    aiclPos++;
    literals++; if(steps) steps.push({type:'literal',char:ch,pos:i});
    i++;
  }
  const result={output,matches,literals,charsIn:chars.length,charsOut:charCount(output)};
  if(steps) result.steps=steps;
  if(trackMapping) result.rawToAicl=rawToAicl;
  return result;
}
export function encodeToString(text){ return encode(text).output; }

File: aicl/encoder.py
"""Stage 1 dictionary encoder."""

from __future__ import annotations

import re
from typing import Any, Mapping

from .dict import TrieNode, build
from .unicode import ESCAPE_MARKER, code_points, is_pua_code_point, require_text

_WORD_RE = re.compile(r"[A-Za-z0-9_'-]")
_LOWER_RE = re.compile(r"[a-z]")
_UPPER_RE = re.compile(r"[A-Z]")
_PUNCTUATION_MODIFIERS: Mapping[str, str] = {
    " ": "MOD_TRAIL_SPACE",
    ",": "MOD_TRAIL_COMMA",
    ".": "MOD_TRAIL_PERIOD",
    "?": "MOD_TRAIL_QUESTION",
    "!": "MOD_TRAIL_EXCL",
    ";": "MOD_TRAIL_SEMI",
    ":": "MOD_TRAIL_COLON",
    ")": "MOD_TRAIL_RPAREN",
    "]": "MOD_TRAIL_RBRACKET",
    "}": "MOD_TRAIL_RBRACE",
    '"': "MOD_TRAIL_RQUOTE",
}


def _trie_match(root: TrieNode, chars: list[str], start: int) -> tuple[int, str | None]:
    node = root
    best_length = 0
    best_symbol: str | None = None
    for end in range(start, len(chars)):
        node = node.children.get(chars[end])  # type: ignore[assignment]
        if node is None:
            break
        if node.symbol is not None:
            best_length = end - start + 1
            best_symbol = node.symbol
    return best_length, best_symbol


def _extract_word(chars: list[str], start: int) -> str:
    if start >= len(chars) or _WORD_RE.fullmatch(chars[start]) is None:
        return ""
    end = start + 1
    while end < len(chars) and _WORD_RE.fullmatch(chars[end]) is not None:
        end += 1
    return "".join(chars[start:end])


def _punctuation_modifier(ch: str) -> str | None:
    return _PUNCTUATION_MODIFIERS.get(ch)


def _all_caps(word: str) -> bool:
    return word == word.upper() and _UPPER_RE.search(word) is not None


def encode(text: str, opts: Mapping[str, Any] | None = None) -> dict[str, Any]:
    """Encode *text* into the stable AICL representation.

    Result keys intentionally use the JavaScript spelling (``charsIn``,
    ``rawToAicl``, etc.) because those objects cross the Python/browser boundary.
    """

    require_text(text, "text")
    options = opts or {}
    steps_enabled = bool(options.get("steps", options.get("stepsEnabled", False)))
    track_mapping = bool(options.get("trackMapping", options.get("track_mapping", False)))
    steps: list[dict[str, Any]] | None = [] if steps_enabled else None

    data = build()
    pattern_to_symbol: Mapping[str, str] = data["patternToSymbol"]  # type: ignore[assignment]
    trie: TrieNode = data["trie"]  # type: ignore[assignment]
    fragment_patterns: tuple[str, ...] = data["fragmentPatterns"]  # type: ignore[assignment]

    chars = code_points(text)
    raw_to_aicl = [-1] * len(chars) if track_mapping else None
    output: list[str] = []
    matches = 0
    literals = 0
    aicl_pos = 0
    index = 0

    while index < len(chars):
        best_length, best_symbol = _trie_match(trie, chars, index)
        best_pattern = "".join(chars[index : index + best_length]) if best_symbol else None

        if best_length == 1:
            word = _extract_word(chars, index)
            if len(word) > 1:
                lower = word.lower()
                is_all_caps = _all_caps(word)
                is_all_lower = word == lower
                is_title_case = (
                    not is_all_lower
                    and word[0] != lower[0]
                    and word[1:] == lower[1:]
                )
                if is_all_caps:
                    caps_symbol = pattern_to_symbol.get(word)
                    if caps_symbol is not None:
                        if raw_to_aicl is not None:
                            for raw_index in range(index, index + len(word)):
                                raw_to_aicl[raw_index] = aicl_pos
                        output.append(caps_symbol)
                        matches += 1
                        if steps is not None:
                            steps.append({"type": "base", "pattern": word, "pos": index})
                        aicl_pos += 1
                        index += len(word)
                        index, matches, aicl_pos = _consume_trailing_modifiers(
                            chars, index, output, pattern_to_symbol, matches, steps, aicl_pos
                        )
                        continue

                    base_symbol = pattern_to_symbol.get(lower)
                    if base_symbol is not None:
                        if raw_to_aicl is not None:
                            for raw_index in range(index, index + len(word)):
                                raw_to_aicl[raw_index] = aicl_pos
                        output.append(base_symbol)
                        matches += 1
                        if steps is not None:
                            steps.append({"type": "base", "pattern": lower, "pos": index})
                        aicl_pos += 1
                        all_caps_symbol = pattern_to_symbol.get("MOD_ALLCAPS")
                        if all_caps_symbol is not None:
                            output.append(all_caps_symbol)
                            matches += 1
                            if steps is not None:
                                steps.append({"type": "modifier", "name": "MOD_ALLCAPS", "pos": index})
                            aicl_pos += 1
                        index += len(word)
                        index, matches, aicl_pos = _consume_trailing_modifiers(
                            chars, index, output, pattern_to_symbol, matches, steps, aicl_pos
                        )
                        continue
                else:
                    if is_all_lower or is_title_case:
                        base_symbol = pattern_to_symbol.get(lower)
                        if base_symbol is not None:
                            if raw_to_aicl is not None:
                                for raw_index in range(index, index + len(lower)):
                                    raw_to_aicl[raw_index] = aicl_pos
                            output.append(base_symbol)
                            matches += 1
                            if steps is not None:
                                steps.append({"type": "base", "pattern": lower, "pos": index})
                            aicl_pos += 1
                            index += len(lower)
                            if word[0] != lower[0]:
                                caps_symbol = pattern_to_symbol.get("MOD_CAPS")
                                if caps_symbol is not None:
                                    output.append(caps_symbol)
                                    matches += 1
                                    if steps is not None:
                                        steps.append({"type": "modifier", "name": "MOD_CAPS", "pos": index})
                                    aicl_pos += 1
                            index, matches, aicl_pos = _consume_trailing_modifiers(
                                chars, index, output, pattern_to_symbol, matches, steps, aicl_pos
                            )
                            continue

                fragment_index = 0
                matched_fragment = False
                while is_all_lower and fragment_index < len(lower):
                    best_fragment: str | None = None
                    best_fragment_length = 0
                    for fragment in fragment_patterns:
                        if len(fragment) <= len(lower) - fragment_index and lower.startswith(
                            fragment, fragment_index
                        ):
                            # All-caps words are fixed by MOD_ALLCAPS.  Other
                            # fragments may use MOD_CAPS only when all interior
                            # characters are lowercase.
                            if is_all_caps or word[fragment_index + 1 : fragment_index + len(fragment)] == lower[
                                fragment_index + 1 : fragment_index + len(fragment)
                            ]:
                                best_fragment = fragment
                                best_fragment_length = len(fragment)
                                break
                    if best_fragment is not None:
                        if raw_to_aicl is not None:
                            for raw_index in range(
                                index + fragment_index,
                                index + fragment_index + best_fragment_length,
                            ):
                                raw_to_aicl[raw_index] = aicl_pos
                        fragment_symbol = pattern_to_symbol.get(best_fragment)
                        if fragment_symbol is not None:
                            output.append(fragment_symbol)
                            matches += 1
                            if steps is not None:
                                steps.append(
                                    {
                                        "type": "fragment",
                                        "pattern": best_fragment,
                                        "pos": index + fragment_index,
                                    }
                                )
                            aicl_pos += 1
                            needs_caps = (
                                _LOWER_RE.search(best_fragment) is not None
                                if is_all_caps
                                else word[fragment_index] != lower[fragment_index]
                            )
                            if needs_caps:
                                caps_symbol = pattern_to_symbol.get(
                                    "MOD_ALLCAPS" if is_all_caps else "MOD_CAPS"
                                )
                                if caps_symbol is not None:
                                    output.append(caps_symbol)
                                    matches += 1
                                    if steps is not None:
                                        steps.append(
                                            {
                                                "type": "modifier",
                                                "name": "MOD_ALLCAPS" if is_all_caps else "MOD_CAPS",
                                                "pos": index + fragment_index,
                                            }
                                        )
                                    aicl_pos += 1
                            fragment_index += best_fragment_length
                            matched_fragment = True
                            continue
                    break
                if matched_fragment and fragment_index > 0:
                    index += fragment_index
                    index, matches, aicl_pos = _consume_trailing_modifiers(
                        chars, index, output, pattern_to_symbol, matches, steps, aicl_pos
                    )
                    continue

        if best_symbol is not None:
            if raw_to_aicl is not None:
                for raw_index in range(index, index + best_length):
                    raw_to_aicl[raw_index] = aicl_pos
            output.append(best_symbol)
            matches += 1
            if steps is not None:
                steps.append({"type": "match", "pattern": best_pattern, "pos": index})
            aicl_pos += len(best_symbol) if best_length > 1 else 1
            index += best_length
            continue

        ch = chars[index]
        if raw_to_aicl is not None:
            raw_to_aicl[index] = aicl_pos
        if is_pua_code_point(ord(ch)):
            output.append(ESCAPE_MARKER)
            output.append(ch)
        else:
            output.append(ch)
        aicl_pos += 1
        literals += 1
        if steps is not None:
            steps.append({"type": "literal", "char": ch, "pos": index})
        index += 1

    result: dict[str, Any] = {
        "output": "".join(output),
        "matches": matches,
        "literals": literals,
        "charsIn": len(chars),
        "charsOut": len("".join(output)),
    }
    if steps is not None:
        result["steps"] = steps
    if raw_to_aicl is not None:
        result["rawToAicl"] = raw_to_aicl
    return result


def _consume_trailing_modifiers(
    chars: list[str],
    index: int,
    output: list[str],
    pattern_to_symbol: Mapping[str, str],
    matches: int,
    steps: list[dict[str, Any]] | None,
    aicl_pos: int,
) -> tuple[int, int, int]:
    """Consume the same run of punctuation modifiers as the JS encoder."""

    while index < len(chars):
        modifier_name = _punctuation_modifier(chars[index])
        if modifier_name is not None:
            modifier_symbol = pattern_to_symbol.get(modifier_name)
            if modifier_symbol is not None:
                output.append(modifier_symbol)
                matches += 1
                aicl_pos += len(modifier_symbol)
                if steps is not None:
                    steps.append({"type": "modifier", "name": modifier_name, "pos": index})
                index += 1
                continue
        break
    return index, matches, aicl_pos


def encode_to_string(text: str) -> str:
    return encode(text)["output"]


# JavaScript-shaped aliases.
encodeToString = encode_to_string

__all__ = ["encode", "encodeToString", "encode_to_string"]
